File: native/android/app/src/main/java/jp/pararia/teacherapp/domain/TeacherModels.kt
package jp.pararia.teacherapp.domain

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class TeacherClientPlatform {
    IOS,
    ANDROID,
    WEB,
    UNKNOWN,
}

@Serializable
data class TeacherClientInfo(
    val platform: TeacherClientPlatform,
    val appVersion: String? = null,
    val buildNumber: String? = null,
)

@Serializable
data class TeacherSession(
    val userId: String,
    val organizationId: String,
    val deviceId: String,
    val role: String,
    val roleLabel: String,
    val userName: String? = null,
    val userEmail: String? = null,
    val deviceLabel: String,
    val issuedAt: String,
    val expiresAt: String,
)

@Serializable
data class TeacherAuthBundle(
    val accessToken: String,
    val accessTokenExpiresAt: String,
    val refreshToken: String,
    val refreshTokenExpiresAt: String,
    val authSessionId: String,
    val tokenType: String,
)

@Serializable
data class TeacherNativeAuthResponse(
    val session: TeacherSession,
    val client: TeacherClientInfo,
    val auth: TeacherAuthBundle,
)

@Serializable
data class TeacherSessionEnvelope(
    val session: TeacherSession,
)

@Serializable
data class TeacherLogoutResponse(
    val ok: Boolean,
)

@Serializable
data class TeacherStudentCandidate(
    val id: String,
    val name: String,
    val subtitle: String? = null,
    val score: Double? = null,
    val reason: String? = null,
)

@Serializable
data class TeacherStudentSearchEnvelope(
    val students: List<TeacherStudentCandidate> = emptyList(),
)

@Serializable
enum class TeacherRecordingStatus {
    @SerialName("RECORDING")
    RECORDING,

    @SerialName("TRANSCRIBING")
    TRANSCRIBING,

    @SerialName("AWAITING_STUDENT_CONFIRMATION")
    AWAITING_STUDENT_CONFIRMATION,

    @SerialName("STUDENT_CONFIRMED")
    STUDENT_CONFIRMED,

    @SerialName("CANCELLED")
    CANCELLED,

    @SerialName("ERROR")
    ERROR,
}

@Serializable
data class TeacherRecordingSummary(
    val id: String,
    val status: TeacherRecordingStatus,
    val deviceLabel: String,
    val recordedAt: String? = null,
    val uploadedAt: String? = null,
    val analyzedAt: String? = null,
    val confirmedAt: String? = null,
    val durationSeconds: Double? = null,
    val transcriptText: String? = null,
    val candidates: List<TeacherStudentCandidate> = emptyList(),
    val errorMessage: String? = null,
)

@Serializable
data class TeacherRecordingEnvelope(
    val recording: TeacherRecordingSummary? = null,
)

@Serializable
data class TeacherActiveRecordingEnvelope(
    val activeRecording: TeacherRecordingSummary? = null,
)

@Serializable
data class TeacherCreateRecordingResponse(
    val recordingId: String,
)

@Serializable
data class TeacherConfirmRecordingResponse(
    val ok: Boolean,
)

@Serializable
data class PendingUpload(
    val id: String,
    val recordingId: String,
    val filePath: String,
    val createdAt: String,
    val durationSeconds: Double? = null,
    val attemptCount: Int = 0,
    val lastAttemptAt: String? = null,
    val errorMessage: String? = null,
)

@Serializable
data class TeacherPersistentState(
    val authBundle: TeacherAuthBundle? = null,
    val pendingUploads: List<PendingUpload> = emptyList(),
)

data class DeviceLoginInput(
    val email: String,
    val password: String,
    val deviceLabel: String,
)

enum class RecorderPermissionStatus {
    UNDETERMINED,
    GRANTED,
    DENIED,
}

enum class RecorderStartAvailability {
    AVAILABLE,
    SYSTEM_AUDIO_BUSY,
}

sealed interface TeacherRoute {
    data object Bootstrap : TeacherRoute
    data object Standby : TeacherRoute
    data class Recording(val seconds: Int, val paused: Boolean = false) : TeacherRoute
    data class Analyzing(val recordingId: String, val message: String) : TeacherRoute
    data class Confirm(val summary: TeacherRecordingSummary) : TeacherRoute
    data class ManualStudentSelect(
        val summary: TeacherRecordingSummary,
        val query: String = "",
        val results: List<TeacherStudentCandidate> = emptyList(),
    ) : TeacherRoute
    data class Done(val title: String, val message: String) : TeacherRoute
    data object Pending : TeacherRoute
}

data class TeacherUiState(
    val session: TeacherSession? = null,
    val route: TeacherRoute = TeacherRoute.Bootstrap,
    val pendingUploads: List<PendingUpload> = emptyList(),
    val errorMessage: String? = null,
    val diagnosticReportText: String = "",
    val requestMicrophonePermission: Boolean = false,
    val requestNotificationPermission: Boolean = false,
)

data class CompletedRecording(
    val filePath: String,
    val durationSeconds: Double,
)

File: native/ios/TeacherNativeApp/Domain/TeacherModels.swift
import Foundation

enum TeacherPlatform: String, Codable {
    case ios = "IOS"
}

struct TeacherClientInfo: Codable {
    let platform: String
    let appVersion: String?
    let buildNumber: String?

    static let current = TeacherClientInfo(
        platform: TeacherPlatform.ios.rawValue,
        appVersion: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String,
        buildNumber: Bundle.main.infoDictionary?["CFBundleVersion"] as? String
    )
}

struct TeacherSession: Codable {
    let userId: String
    let organizationId: String
    let deviceId: String
    let role: String
    let roleLabel: String
    let userName: String?
    let userEmail: String?
    let deviceLabel: String
    let issuedAt: String
    let expiresAt: String
}

struct TeacherAuthBundle: Codable {
    let accessToken: String
    let accessTokenExpiresAt: String
    let refreshToken: String
    let refreshTokenExpiresAt: String
    let authSessionId: String
    let tokenType: String
}

struct TeacherNativeAuthResponse: Codable {
    let session: TeacherSession
    let client: TeacherClientInfo
    let auth: TeacherAuthBundle
}

struct TeacherSessionEnvelope: Codable {
    let session: TeacherSession
}

struct TeacherLogoutResponse: Codable {
    let ok: Bool
}

struct TeacherRecordingEnvelope: Codable {
    let recording: TeacherRecordingSummary?
}

struct TeacherCreateRecordingResponse: Codable {
    let recordingId: String
}

struct TeacherConfirmRecordingResponse: Codable {
    let ok: Bool
}

enum TeacherRecordingStatus: String, Codable {
    case recording = "RECORDING"
    case transcribing = "TRANSCRIBING"
    case awaitingStudentConfirmation = "AWAITING_STUDENT_CONFIRMATION"
    case studentConfirmed = "STUDENT_CONFIRMED"
    case cancelled = "CANCELLED"
    case error = "ERROR"
}

struct TeacherStudentCandidate: Codable, Identifiable, Hashable {
    let id: String
    let name: String
    let subtitle: String?
    let score: Double?
    let reason: String?
}

struct TeacherRecordingSummary: Codable, Identifiable {
    let id: String
    let status: TeacherRecordingStatus
    let deviceLabel: String
    let recordedAt: String?
    let uploadedAt: String?
    let analyzedAt: String?
    let confirmedAt: String?
    let durationSeconds: Double?
    let transcriptText: String?
    let candidates: [TeacherStudentCandidate]
    let errorMessage: String?
}

struct PendingUpload: Codable, Identifiable, Hashable {
    let id: String
    let recordingId: String
    let fileURL: URL
    let createdAt: Date
    let durationSeconds: Double?
    let attemptCount: Int
    let lastAttemptAt: Date?
    var errorMessage: String?

    init(
        id: String,
        recordingId: String,
        fileURL: URL,
        createdAt: Date,
        durationSeconds: Double? = nil,
        attemptCount: Int = 0,
        lastAttemptAt: Date? = nil,
        errorMessage: String? = nil
    ) {
        self.id = id
        self.recordingId = recordingId
        self.fileURL = fileURL
        self.createdAt = createdAt
        self.durationSeconds = durationSeconds
        self.attemptCount = attemptCount
        self.lastAttemptAt = lastAttemptAt
        self.errorMessage = errorMessage
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case recordingId
        case fileURL
        case createdAt
        case durationSeconds
        case attemptCount
        case lastAttemptAt
        case errorMessage
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        recordingId = try container.decode(String.self, forKey: .recordingId)
        fileURL = try container.decode(URL.self, forKey: .fileURL)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        durationSeconds = try container.decodeIfPresent(Double.self, forKey: .durationSeconds)
        attemptCount = try container.decodeIfPresent(Int.self, forKey: .attemptCount) ?? 0
        lastAttemptAt = try container.decodeIfPresent(Date.self, forKey: .lastAttemptAt)
        errorMessage = try container.decodeIfPresent(String.self, forKey: .errorMessage)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(recordingId, forKey: .recordingId)
        try container.encode(fileURL, forKey: .fileURL)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encodeIfPresent(durationSeconds, forKey: .durationSeconds)
        try container.encode(attemptCount, forKey: .attemptCount)
        try container.encodeIfPresent(lastAttemptAt, forKey: .lastAttemptAt)
        try container.encodeIfPresent(errorMessage, forKey: .errorMessage)
    }
}

struct DeviceLoginInput {
    let email: String
    let password: String
    let deviceLabel: String
}

enum RecorderPermissionStatus {
    case undetermined
    case granted
    case denied
}

enum TeacherRoute: Equatable {
    case bootstrap
    case standby
    case recording(seconds: Int)
    case analyzing(recordingId: String, message: String)
    case confirm(TeacherRecordingSummary)
    case done(title: String, message: String)
    case pending
}

protocol TeacherTokenStore {
    func loadAuthBundle() -> TeacherAuthBundle?
    func save(authBundle: TeacherAuthBundle) throws
    func clear() throws
}

protocol TeacherAuthRepository {
    func currentSession() async -> TeacherSession?
    func login(input: DeviceLoginInput) async throws -> TeacherSession
    func refreshIfNeeded() async throws -> TeacherSession?
    func forceRefresh() async throws -> TeacherSession?
    func logout() async
}

protocol TeacherRecordingRepository {
    func loadActiveRecording() async throws -> TeacherRecordingSummary?
    func createRecording() async throws -> String
    func uploadAudio(recordingId: String, fileURL: URL, durationSeconds: Double?) async throws -> TeacherRecordingSummary?
    func pollRecording(recordingId: String) async throws -> TeacherRecordingSummary
    func confirmStudent(recordingId: String, studentId: String?) async throws
    func cancelRecording(recordingId: String) async throws
    func retryPendingUploads() async throws
}

protocol PendingUploadStore {
    func loadItems() throws -> [PendingUpload]
    func save(_ item: PendingUpload) throws
    func remove(id: String) throws
}

protocol AudioRecorderClient {
    func permissionStatus() async -> RecorderPermissionStatus
    func requestPermission() async -> RecorderPermissionStatus
    func start() throws
    func stop() throws -> URL
    func cancel() throws
}
